src/gcd/binary.rs

```diff
@@ -0,0 +1,305 @@
+    let mut final_size = result_cur.len() + shift_words;
+    for i in 0..shift_words {
+        // LEGACY: equivalent to lhs[..shift_words].fill(0) after Rust 1.50
+        lhs[i] = 0;
+    }
+    lhs[shift_words..final_size].copy_from_slice(result_cur);
+    let carry = shift::shl_in_place(
+        &mut lhs[shift_words..final_size],
+        (init_zeros % WORD_BITS_USIZE) as u32,
+    );
+    if carry > 0 {
+        lhs[final_size] = carry;
+        final_size += 1;
+    }
+    final_size
+}
+
+/// Temporary memory required for extended gcd.
+pub(crate) fn memory_requirement_up_to(lhs_len: usize, rhs_len: usize) -> Layout {
+    // Required memory:
+    // - two numbers (s0 & s1) with at most the same size as rhs
+    // - two numbers (t0 & t1) with at most the same size as lhs
+    let num_words = 2 * (lhs_len + rhs_len);
+    memory::array_layout::<Word>(num_words)
+}
+
+/// Extended binary GCD for two multi-digits numbers
+pub(crate) fn xgcd_in_place(
+    lhs: &mut [Word],
+    rhs: &mut [Word],
+    g: &mut [Word],
+    bonly: bool,
+    memory: &mut Memory,
+) -> (Sign, Sign) {
+    debug_assert!(lhs.len() > 1 && rhs.len() > 1);
+
+    // find common factors of 2
+    let lhs_zeros = trailing_zeros_large(lhs);
+    let rhs_zeros = trailing_zeros_large(rhs);
+
+    let (lhs_pos, rhs_pos) = (lhs_zeros / WORD_BITS_USIZE, rhs_zeros / WORD_BITS_USIZE);
+    shift::shr_in_place(&mut lhs[lhs_pos..], (lhs_zeros % WORD_BITS_USIZE) as u32);
+    shift::shr_in_place(&mut rhs[rhs_pos..], (rhs_zeros % WORD_BITS_USIZE) as u32);
+
+    // allocate memory for coefficients
+    let (s0, mut memory) = memory.allocate_slice_fill::<Word>(lhs.len(), 0);
+    let (s1, mut memory) = memory.allocate_slice_fill::<Word>(lhs.len(), 0);
+    let (t0, mut memory) = memory.allocate_slice_fill::<Word>(rhs.len(), 0);
+    let (t1, _) = memory.allocate_slice_fill::<Word>(rhs.len(), 0);
+    let (mut s0_end, mut s1_end) = (1, 1);
+    let (mut t0_end, mut t1_end) = (1, 1);
+
+    // initialize s, t
+    let (init_zeros, mut shift) = match lhs_zeros.cmp(&rhs_zeros) {
+        Ordering::Equal => {
+            *s0.first_mut().unwrap() = 1;
+            *t1.first_mut().unwrap() = 1;
+            (lhs_zeros, 0)
+        }
+        Ordering::Greater => {
+            let shift = lhs_zeros - rhs_zeros;
+            *s0.first_mut().unwrap() = 1;
+            *t1.get_mut(shift / WORD_BITS_USIZE).unwrap() |= 1 << (shift % WORD_BITS_USIZE);
+            t1_end += shift / WORD_BITS_USIZE;
+            (rhs_zeros, shift)
+        }
+        Ordering::Less => {
+            let shift = rhs_zeros - lhs_zeros;
+            *s0.get_mut(shift / WORD_BITS_USIZE).unwrap() |= 1 << (shift % WORD_BITS_USIZE);
+            s0_end += shift / WORD_BITS_USIZE;
+            *t1.first_mut().unwrap() = 1;
+            (lhs_zeros, shift)
+        }
+    };
+
+    /// lhs[..lhs_end] += rhs, assuming rhs.len() <= lhs.len()
+    /// return true if lhs is fully overflowed
+    #[inline(always)]
```

Comment:
Why inline(always)? No need for this, let the compiler decide.

Reply:
Done in https://github.com/tczajka/ibig-rs/pull/11/commits/a66b5164f35520c67b6bf825044f750537690e88